pkg/reconciler/build/resources/pod.go

```diff
@@ -491,3 +525,470 @@ func getFailureMessage(pod *corev1.Pod) string {
 	// Lastly fall back on a generic error message.
 	return "build failed for unspecified reasons."
 }
+
+// from build-pipeline
+
+const (
+	// MountName is the name of the pvc being mounted (which
+	// will contain the entrypoint binary and eventually the logs)
+	MountName         = "tools"
+	MountPoint        = "/tools"
+	BinaryLocation    = MountPoint + "/entrypoint"
+	JSONConfigEnvVar  = "ENTRYPOINT_OPTIONS"
+	InitContainerName = "place-tools"
+	// TODO(aaron-prindle) change this to wherever is sensible
+	DefaultEntrypointImage = "gcr.io/aprindle-vm-test/entrypoint:latest"
+
+	ProcessLogFile        = "/tools/process-log.txt"
+	MarkerFile            = "/tools/marker-file.txt"
+	ShouldWaitForPrevStep = false
+	PreRunFile            = "0"
+	ShouldRunPostRun      = true
+	PostRunFile           = "0"
+)
+
+var toolsMount = corev1.VolumeMount{
+	Name:      MountName,
+	MountPath: MountPoint,
+}
+
+var toolsVolume = corev1.Volume{
+	Name: MountName,
+	VolumeSource: corev1.VolumeSource{
+		EmptyDir: &corev1.EmptyDirVolumeSource{},
+	},
+}
+
+type entrypointArgs struct {
+	Args       []string `json:"args"`
+	ProcessLog string   `json:"process_log"`
+	MarkerFile string   `json:"marker_file"`
+
+	ShouldWaitForPrevStep bool   `json:"shouldWaitForPrevStep"`
+	PreRunFile            string `json:"preRunFile"`
+	ShouldRunPostRun      bool   `json:"shouldRunPostRun"`
+	PostRunFile           string `json:"postRunFile"`
+}
+
+// Cache is a simple caching mechanism allowing for caching the results of
+// getting the Entrypoint of a container image from a remote registry. It
+// is synchronized via a mutex so that we can share a single Cache across
+// each worker thread that the reconciler is running. The mutex is necessary
+// due to the possibility of a panic if two workers were to attempt to read and
+// write to the internal map at the same time.
+type Cache struct {
+	mtx   sync.RWMutex
+	cache map[string][]string
+}
+
+// NewCache is a simple helper function that returns a pointer to a Cache that
+// has had the internal cache map initialized.
+func NewCache() *Cache {
+	return &Cache{
+		cache: make(map[string][]string),
+	}
+}
+
+func (c *Cache) get(sha string) ([]string, bool) {
+	c.mtx.RLock()
+	ep, ok := c.cache[sha]
+	c.mtx.RUnlock()
+	return ep, ok
+}
+
+func (c *Cache) set(sha string, ep []string) {
+	c.mtx.Lock()
+	c.cache[sha] = ep
+	c.mtx.Unlock()
+}
+
+type authToken struct {
```

Comment:
Mentioned this before, but I think like 95% of the added code in this file could go away if you do something like:

```go
kc, _ := k8schain.New(kubeclient, k8schain.Options{
  ImagePullSecrets: secret.Name,
})
img, _ := remote.Image(ref, remote.WithAuthFromKeychain(kc))

```

Obviously handle errors too, but what you're doing here is basically the point of `k8schain`.

https://github.com/google/go-containerregistry/blob/e0f538712eb0767ae6959531febe6041fa466602/pkg/authn/k8schain/k8schain.go#L46

Reply:
Awesome, this is working for me now! Note: I have to use go-contaierregistry w/ this change: https://github.com/google/go-containerregistry/pull/328 due to some k8s version import mismatch.  More info in that PR